File: integrated-app/integrated-app-gwt-gwt/src/main/java/com/grafapp/model/Graph.java
package com.grafapp.model;

import java.util.*;

/**
 * Model graf inti menggunakan adjacency list.
 * Mendukung graf directed dan undirected, weighted dan unweighted.
 */
public class Graph {
    private final Map<Integer, GraphNode> nodes = new LinkedHashMap<>();
    private final List<GraphEdge> edges = new ArrayList<>();
    private boolean directed;
    private boolean weighted;

    public Graph() { this(false); }

    public Graph(boolean directed) { this.directed = directed; }

    public boolean isWeighted() { return weighted; }
    public void setWeighted(boolean weighted) { this.weighted = weighted; }

    public void addNode(int id) {
        if (!nodes.containsKey(id)) {
            nodes.put(id, new GraphNode(id));
        }
    }

    public void addNode(GraphNode node) {
        nodes.put(node.getId(), node);
    }

    public void addEdge(int source, int target) {
        addNode(source);
        addNode(target);
        for (GraphEdge e : edges) {
            if (e.connects(source, target)) return;
        }
        edges.add(new GraphEdge(source, target));
    }

    public void addEdge(int source, int target, double weight) {
        addNode(source);
        addNode(target);
        for (GraphEdge e : edges) {
            if (e.connects(source, target)) return;
        }
        edges.add(new GraphEdge(source, target, weight));
    }

    public void removeNode(int id) {
        nodes.remove(id);
        edges.removeIf(e -> e.getSource() == id || e.getTarget() == id);
    }

    public void removeEdge(int source, int target) {
        edges.removeIf(e -> e.connects(source, target));
    }

    public GraphNode getNode(int id) { return nodes.get(id); }
    public Collection<GraphNode> getNodes() { return nodes.values(); }
    public List<GraphEdge> getEdges() { return edges; }
    public int getNodeCount() { return nodes.size(); }
    public int getEdgeCount() { return edges.size(); }
    public boolean isDirected() { return directed; }
    public void setDirected(boolean directed) { this.directed = directed; }
    public Set<Integer> getNodeIds() { return nodes.keySet(); }

    public List<Integer> getNeighbors(int nodeId) {
        List<Integer> neighbors = new ArrayList<>();
        for (GraphEdge edge : edges) {
            if (edge.getSource() == nodeId) {
                neighbors.add(edge.getTarget());
            } else if (!directed && edge.getTarget() == nodeId) {
                neighbors.add(edge.getSource());
            }
        }
        Collections.sort(neighbors);
        return neighbors;
    }

    public GraphEdge getEdge(int source, int target) {
        for (GraphEdge e : edges) {
            if (directed) {
                if (e.getSource() == source && e.getTarget() == target) return e;
            } else {
                if (e.connects(source, target)) return e;
            }
        }
        return null;
    }

    public void resetStates() {
        for (GraphNode node : nodes.values()) {
            node.setState(NodeState.UNVISITED);
        }
        for (GraphEdge edge : edges) {
            edge.setState(EdgeState.DEFAULT);
        }
    }

    public int nextAvailableId() {
        int max = -1;
        for (int id : nodes.keySet()) {
            if (id > max) max = id;
        }
        return max + 1;
    }

    public void clear() {
        nodes.clear();
        edges.clear();
    }
}

File: web/src/lib/graph/graph.ts
export type NodeStateKey = string;

export type EdgeStateKey =
  | "DEFAULT"
  | "TRAVERSING"
  | "VISITED"
  | "BRIDGE"
  | "PATH"
  | "TREE_EDGE";

export const nodeStateStyles: Record<string, { fill: string; fillLighter: string; stroke: string }> = {
  UNVISITED: { fill: "#ffffff", fillLighter: "#f7f7f7", stroke: "#9e9e9e" },
  PROCESSING: { fill: "#fff9c4", fillLighter: "#fffde7", stroke: "#f9a825" },
  VISITED: { fill: "#bbdefb", fillLighter: "#e3f2fd", stroke: "#1976d2" },
  PATH: { fill: "#c8e6c9", fillLighter: "#e8f5e9", stroke: "#388e3c" },
  START: { fill: "#90caf9", fillLighter: "#e3f2fd", stroke: "#0d47a1" },
  END: { fill: "#ffcdd2", fillLighter: "#ffebee", stroke: "#c62828" },
  ARTICULATION: { fill: "#ffe0b2", fillLighter: "#fff3e0", stroke: "#e65100" }
};

// Generate 30 component styles dynamically with even color distribution (golden angle)
const componentStates: string[] = [];
for (let i = 1; i <= 30; i++) {
  const stateName = `COMPONENT_${i}`;
  componentStates.push(stateName);
  const hue = Math.floor((i * 137.5) % 360);
  nodeStateStyles[stateName] = {
    fill: `hsl(${hue}, 85%, 92%)`,
    fillLighter: `hsl(${hue}, 95%, 97%)`,
    stroke: `hsl(${hue}, 65%, 40%)`
  };
}

export const edgeStateStyles: Record<EdgeStateKey, { color: string; width: number }> = {
  DEFAULT: { color: "#bdbdbd", width: 1.5 },
  TRAVERSING: { color: "#ffc107", width: 3.0 },
  VISITED: { color: "#64b5f6", width: 2.5 },
  BRIDGE: { color: "#f44336", width: 3.5 },
  PATH: { color: "#4caf50", width: 3.0 },
  TREE_EDGE: { color: "#00897b", width: 3.5 }
};

export const componentStateFor = (id: number): string => {
  const index = Math.abs(id) % componentStates.length;
  return componentStates[index];
};

export class GraphNode {
  id: number;
  label: string;
  x = 0;
  y = 0;
  coordinateX = 0;
  coordinateY = 0;
  hasCoordinate = false;
  vx = 0;
  vy = 0;
  fx = 0;
  fy = 0;
  state: NodeStateKey = "UNVISITED";
  pinned = false;

  constructor(id: number, label?: string) {
    this.id = id;
    this.label = label ?? String(id);
  }

  resetForce() {
    this.fx = 0;
    this.fy = 0;
  }
}

export class GraphEdge {
  source: number;
  target: number;
  weight: number;
  state: EdgeStateKey = "DEFAULT";

  constructor(source: number, target: number, weight = 1) {
    this.source = source;
    this.target = target;
    this.weight = weight;
  }

  connects(a: number, b: number) {
    return (this.source === a && this.target === b) || (this.source === b && this.target === a);
  }
}

export class Graph {
  private nodes = new Map<number, GraphNode>();
  private edges: GraphEdge[] = [];
  directed = false;
  weighted = false;

  constructor(directed = false) {
    this.directed = directed;
  }

  get nodeCount() {
    return this.nodes.size;
  }

  get edgeCount() {
    return this.edges.length;
  }

  getNode(id: number) {
    return this.nodes.get(id);
  }

  getNodes() {
    return Array.from(this.nodes.values());
  }

  getEdges() {
    return this.edges;
  }

  getNodeIds() {
    return Array.from(this.nodes.keys());
  }

  addNode(idOrNode: number | GraphNode) {
    if (typeof idOrNode === "number") {
      if (!this.nodes.has(idOrNode)) {
        this.nodes.set(idOrNode, new GraphNode(idOrNode));
      }
      return;
    }
    this.nodes.set(idOrNode.id, idOrNode);
  }

  addEdge(source: number, target: number, weight?: number) {
    this.addNode(source);
    this.addNode(target);
    const hasDuplicate = this.edges.some((edge) =>
      this.directed ? edge.source === source && edge.target === target : edge.connects(source, target)
    );
    if (hasDuplicate) return;
    this.edges.push(new GraphEdge(source, target, weight ?? 1));
  }

  removeNode(id: number) {
    this.nodes.delete(id);
    this.edges = this.edges.filter((edge) => edge.source !== id && edge.target !== id);
  }

  removeEdge(source: number, target: number) {
    this.edges = this.edges.filter((edge) =>
      this.directed ? !(edge.source === source && edge.target === target) : !edge.connects(source, target)
    );
  }

  getNeighbors(nodeId: number) {
    const neighbors: number[] = [];
    for (const edge of this.edges) {
      if (edge.source === nodeId) {
        neighbors.push(edge.target);
      } else if (!this.directed && edge.target === nodeId) {
        neighbors.push(edge.source);
      }
    }
    neighbors.sort((a, b) => a - b);
    return neighbors;
  }

  getEdge(source: number, target: number) {
    for (const edge of this.edges) {
      if (this.directed) {
        if (edge.source === source && edge.target === target) return edge;
      } else if (edge.connects(source, target)) {
        return edge;
      }
    }
    return undefined;
  }

  resetStates() {
    for (const node of this.nodes.values()) {
      node.state = "UNVISITED";
    }
    for (const edge of this.edges) {
      edge.state = "DEFAULT";
    }
  }

  nextAvailableId() {
    let max = -1;
    for (const id of this.nodes.keys()) {
      if (id > max) max = id;
    }
    return max + 1;
  }

  clear() {
    this.nodes.clear();
    this.edges = [];
  }
}
